Core/GameObject: Implement DespawnAtAction

Killyana opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-    Fix the logic used by IsDespawnAtAction for chest and goobers
-    Keep Goobers and chests that are summoned, spawned until the timer expire, like Romantic Basket case

**Target branch(es):** 3.3.5

- [x] 3.3.5
- [ ] master

**Issues addressed:** #15730

**Tests performed:** Yes

**Known issues and TODO list:** (add/remove lines as needed)

- [ ]  This works only If I add: m_respawnTime > 0 :
``` cpp
 if (m_respawnTime > 0 &&  !GetGOInfo()->IsDespawnAtAction())
```
But this will exclude DB spawns, I can't figure out why.

Killyana committed:
Core/GameObject: Implement DespawnAtAction

Killyana committed:
Unneeded change

Killyana committed:
Indentation

Killyana committed:
DB/Creature: Fix some Waypoints

Killyana committed:
DB/Creature: Aldor Spawn Controller

Killyana committed:
DB/Creature: Nova Terra

Killyana committed:
DB/Creature: Little Jimmy Vishas

The event with Renault Mograine and Sally Whitemane and a Captured Critter is not scripted

Killyana committed:
DB/Creature: Event Generator Old Hillsbrad

Killyana committed:
DB: Fix rotation and state for Nether Collector Tube and add Coruu Control Console

Killyana committed:
DB/SAI: Overseer Seylanna

And remove some wrong spawns

Killyana committed:
DB/Creature: Duro Control Console

Killyana committed:
DB/Creature: Death Watcher/Entropic Eye

Killyana committed:
DB/Creature: Kaylaan & Fallen Vindicator

Killyana committed:
DB/Creature: Wrath-Scryer's Felfire

Used in spell=35770 by Wrath-Scryer Soccothrates, the boss has half of his abilities missing or not scripted

Killyana committed:
DB/Quest: Items of Power

Closes #6388

Killyana committed:
Rename file

Killyana committed:
DB/Creature: Shadowmoon Villager

And fix some auras and equipements

Killyana committed:
DB/Creature: Tidewalker Warrior

Killyana committed:
DB/Creature: Warder/Defender Corpses

Ref #21102

Killyana committed:
DB/Creature: Floating Skull

Shauren committed:
Core/GameObjects: Validate rotation quaternion at startup

Ref #23306

jackpoz committed:
Core/SmartAI: Fix SMART_EVENT_SUMMON_DESPAWNED

Fix SMART_EVENT_SUMMON_DESPAWNED not working when event_param1 was set to any value other than 0

jackpoz committed:
3.3.5 gameobject summoner (#23289)

* Scripts/Misc: Change IsSummonedBy(Unit*) to IsSummonedBy(WorldObject*)

* Scripts/Misc: Fix build

* Core/TempSummons: Rename GetSummoner() to GetSummonerUnit()

* Core/TempSummons: Add support to TempSummons::GetSummoner() to return GameObject too

* Fix build

* Core/TempSummons: Allow GameObject to be owner of TempSummon

* Core/TempSummons: Add support to SAI for GameObject owner of TempSummon

* Scripts/Misc: Fix no-pch build

* Core/TempSummons: Implement PR comments

Killyana committed:
DB/Quest: Test Flight: The Zephyrium Capacitorium

It fix also the 3 other Test Flight quests
Closes #17840

Killyana committed:
DB/GameObject: Fix wrong rotations

Closes #23306

Killyana committed:
DB/Misc: Remove some auras related to channel spell and add the proper script to Eclipsion Spellbinder & Eclipsion Archmage

Also fix some cosmetic mounts

Commit:
Scripts/Zul'Gurub 22974 high priestess marli hatching eggs (#23212)

* Issue 22974: Initial bugfix suggestion.

* Fixed allignment.

* Removed unused call.

* Initial spider spawn targeting is done.

* Resolves issue 22874. Initial implementation suggestion.

* Commenting out dead code.

* Reordered private members to meet standards of member initializer list.

* Removing unused parameter and changing data structure from list to vector.

* Changed list to vector in the grid notifier.

* Removing comments and fixing identation.

* Stupid collapsed blocks.

* Scripts/Misc: Change IsSummonedBy(Unit*) to IsSummonedBy(WorldObject*)

* Scripts/Misc: Fix build

* Core/TempSummons: Rename GetSummoner() to GetSummonerUnit()

* Core/TempSummons: Add support to TempSummons::GetSummoner() to return GameObject too

* Fix build

* Core/TempSummons: Allow GameObject to be owner of TempSummon

* Core/TempSummons: Add support to SAI for GameObject owner of TempSummon

* Scripts/Misc: Fix no-pch build

* Cleaned up all the core changes and scripted eggs and spell according to pull request comments.

* Fix no-pch build.

* Fix no-pch build 2.

* Added sql file for the spell and gameobject script.

* Update 2019_05_20_00_world.sql

* Added conditions to spell thanks @xvywh and some minor changes to the code.

* Removed some strange leftovers.

* Removed obsolete brackets.

* Rename 2019_05_20_00_world.sql to 2019_05_24_00_world.sql

Kittnz commented:
Might want to add a random or between option

jackpoz committed:
Core/SmartAI: Implement SMART_EVENT_SUMMONED_UNIT and SMART_EVENT_SUMMON_DESPAWNED for GameObject owners

Jildor committed:
Core/Spells: Fix spells with Health Leech effect, now check damage absorb when calculate Heal (#23312)

Closes: #23311

illfated commented on src/server/game/Entities/GameObject/GameObjectData.h:
```diff
@@ -477,7 +477,7 @@ struct GameObjectTemplate
     }
 
-    bool GetDespawnPossibility() const                      // despawn at targeting of cast?
+    bool CannotBeUsedUnderImmunity() const                      // The object cannot be use under Immunity like divine shield -- to be implemented
```
```suggestion
    bool CannotBeUsedUnderImmunity() const                  // The object cannot be used under Immunity like divine shield -- to be implemented
```

Killyana committed:
DB/Quest: Portals of the Legion

Killyana committed:
DB/Misc: Eclipsions Npcs

Killyana committed:
DB/Creature: Dark Portal Dummy 1.30

Killyana committed:
DB: Typo in the previous commit

Killyana committed:
DB/Creature: Hydross the unstable npc triggers

Update the correct channel spell SPELL_BLUE_BEAM and spawn triggers used to cast it

Killyana committed:
DB: Fix wrong spawnMask

Kittnz committed:
Core/Weather: Define WEATHER_STATE_DRIZZLE state

jackpoz committed:
Core/SmartAI: Improve SMART_EVENT_GOSSIP_HELLO

Add another value to event_param1 for SMART_EVENT_GOSSIP_HELLO to support all cases of OnGossipHello/OnReportUse in GameObjects:
- event_param1 set to 0: execute the action for both OnGossipHello and OnReportUse. This might result in the action being executed twice when clicking the GameObject
- event_param1 set to 1: execute the action for only OnGossipHello
- event_param1 set to 2: execute the action for only OnReportUse

Killyana committed:
DB/Quest: Improvements for Death From Below

Killyana committed:
DB/Spell: D.I.S.C.O.

Killyana committed:
DB/SAI: Update gossip hello event for game objects to trigger only in  OnReportUse() and not OnGossipHello()

The only remaining case is the gob 188596 as apparently it triggers OnReportUse() only

sirikfoll committed:
DB/Quest: No Place to Run

Closes #22719

Ovahlord committed:
Core/GameObjects: goober gameobjects may reset only if they have a lock id or a reset time specified (#23298)

* Core/GameObjects: goober gameobjects may reset only if they have a lock id

Tests have shown that Goobers without a lock id (Data0=0) are not allowed to reset their go state such as the teleporters in Ulduar and Icecrown Citadel. The tests has been expanded by checking 4.x goobers as well and the perfect example that confirms that result is the Ancient Bell for Atramedes' intro which also is not suposed to reset after using it.

* Core/GameObjects: allow to reset goobers without a lock id if a reset time is defined

* fixed a typo

Wyrserth committed:
Core/Console: start CliRunnable thread after all log messages have been printed. (#23321)

Killyana committed:
DB/Creature: Restore Bath'rah the Windwatcher script

Closes #23318

Killyana committed:
DB/Quest: Vyletongue Corruption

Ref #21078

lactoseti committed:
Creature/Script: npc_tallhorn_stag

Ref #2623

Ovahlord committed:
Core/Players: out of range group updates will now get sent every 5 seconds instead of as soon as they are available to reflect retail behaivior. (#23323)

* according to sniffs the group update packet for out of range members has a fixed 5 seconds interval instead of spaming packets like crazy for example when updating positions. On 4.x and above this will fix a very nasty and ugly fps drop
* validated with 4.3.4 and 3.3.5 sniffs

Killyana committed:
DB/Creature: Enslaved Doomguard

Killyana committed:
DB/Creature: Dragonmaw Nether Drake

Killyana committed:
DB/Creature: Eclipsion Hawkstrider

Killyana committed:
DB/Creature: Arcane Burst

jackpoz committed:
Core/Movement: Fix assertion triggered in SplineChainMovementGenerator

Fix an assertion triggered in SplineChainMovementGenerator happening when a new generator is added while last spline of the SplineChainMovementGenerator is being processed but is not finalized yet.
Fix #23077

Killyana committed:
DB/Creature: Ethereum Avenger

Killyana committed:
DB/Creature: Outland Children's Week Dark Portal Trigger

dr-j committed:
DB/Creature: Wolver Orphan & Oracle Orphan event in Dalaran

Closes #14621

Killyana committed:
DB/Creature: Black Temple Battle Sensor

Killyana committed:
DB/Creature: Lightsworn Elekk Rider

Killyana committed:
DB/Creature: Add Grip immunity to Nightbane

Killyana committed:
DB/Creature: Scryer Cavalier

Killyana committed:
DB/Quest: The Wrath of Neptulon

jackpoz committed:
Core/Channel:  Fix CRT Debug Assert triggered with invalid channel names

jackpoz committed:
Core/Channel: Sanitize channel names

Wyrserth committed:
Core/Script: add script hook to allow overriding of a vehicle passenger's exit position (#23322)

* Core/Script: add script hook to allow overriding of a vehicle passenger's exit position.

* Allow to change orientation, and fix nopch.

* Keep original orientation, thanks jackpoz!

* Update Unit.cpp

Code cleanup

* Update Unit.cpp

jackpoz committed:
Core/MovementGenerator: Add more details to asserts

Implement GetDebugInfo in MovementGenerator, PathMovementBase, FlightPathMovementGenerator, WaypointMovementGenerator.
Add an additional assert in FlightPathMovementGenerator::DoEventIfAny(), the input node must not be null

Wyrserth committed:
Core/Misc: don't allow players to use/activate/loot non-allowed gameobjects while under the effect of a damage immunity aura (#23319)

* Core/Misc: don't allow players to use/activate/loot non-allowed gameobjects while under the effect of a damage immunity aura.

* Forgot to change this in last-second rename.

* Apply suggested changes, thanks Shauren!

jackpoz commented:
Need help rebasing ?

illfated commented:
Looks like it. Second time in a short row.

Killyana commented:
I don't know how to fix the remaining issue, no need to have the PR open for years.

jackpoz commented:
I can debug it when I have time if you need, just let me know

Killyana commented:
It will be nice.
